File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.js
class Solution {
    maxFreq(map){
        let max = 0;
        for(const [key, value] of map){
            max = Math.max(max, value);
        }
        return max;
    }
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {
        const map = new Map();
        let start = 0, maxLength = 0;

        for(let end = 0; end < s.length; end++){
            const target = s[end];

            map.set(target, (map.get(target) || 0) + 1);
            const freq = this.maxFreq(map);

            if(end - start - freq + 1 <= k){
                maxLength = Math.max(maxLength, end - start + 1);
            } else {
                // reduce the frequency of the outgoing
                const outFreq = map.get(s[start]);
                if(outFreq){
                    map.set(s[start], outFreq - 1);
                }
                start = start + 1;
            }
        }

        return maxLength;
    }
}

File: Data Structures & Algorithms/is-palindrome/submission-1.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        start = 0
        end = len(s) - 1

        while start < end:
            while start < len(s) and not s[start].isalnum():
                start = start + 1
            
            while end >= 0 and not s[end].isalnum():
                end = end - 1
            
            if start >= len(s) or end < 0:
                break
            
            if s[start].lower() != s[end].lower():
                return False
            
            start = start + 1
            end = end - 1
        
        return True